pkg/vault/contracts/BalancerContractRegistry.sol

```diff
@@ -0,0 +1,201 @@
+     * );
+     * registerBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool', 0x201efd508c8DfE9DE1a13c2452863A78CB2a86Cc);
+     *
+     * // Register the Routers (two of them anyway).
+     * registerBalancerContract(ContractType.ROUTER, '20241205-v3-router', 0x5C6fb490BDFD3246EB0bB062c168DeCAF4bD9FDd);
+     * registerBalancerContract(
+     *      ContractType.ROUTER, '20241205-v3-batch-router', 0x136f1EFcC3f8f88516B9E94110D56FDBfB1778d1
+     * );
+     *
+     * // Now, hooks that require trusted routers can be deployed with the registry address, and query the router to
+     * // see whether it's "trusted" (i.e., registered by governance):
+     *
+     * isActiveBalancerContract(ContractType.ROUTER, 0x5C6fb490BDFD3246EB0bB062c168DeCAF4bD9FDd) would return true.
+     *
+     * Off-chain processes that wanted to know the current address of the Weighted Pool Factory could query by either
+     * name:
+     *
+     * (address, active) = getBalancerContract(ContractType.POOL_FACTORY, '20241205-v3-weighted-pool');
+     * (address, active) = getBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool');
+     *
+     * These would return the same result.
+     *
+     * If we replaced `20241205-v3-weighted-pool` with `20250107-v3-weighted-pool-v2`, governance would call:
+     *
+     * deprecateBalancerContract(0x201efd508c8DfE9DE1a13c2452863A78CB2a86Cc);
+     * registerBalancerContract(
+     *      ContractType.POOL_FACTORY, '20250107-v3-weighted-pool-v2', 0x9FC3da866e7DF3a1c57adE1a97c9f00a70f010c8)
+     * );
+     * replaceBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool', 0x9FC3da866e7DF3a1c57adE1a97c9f00a70f010c8);
+     *
+     * At that point,
+     * getBalancerContract(ContractType.POOL_FACTORY, '20241205-v3-weighted-pool') returns active=false,
+     * isActiveBalancerContract(ContractType.POOL_FACTORY, 0x201efd508c8DfE9DE1a13c2452863A78CB2a86Cc) returns false,
+     * getBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool') returns the v2 address (and active=true).
+     */
+
+    /// @inheritdoc IBalancerContractRegistry
+    function registerBalancerContract(
+        ContractType contractType,
+        string memory contractName,
+        address contractAddress
+    ) external authenticate {
+        if (contractAddress == address(0)) {
+            revert ZeroContractAddress();
+        }
+
+        if (bytes(contractName).length == 0) {
+            revert InvalidContractName();
+        }
+
+        bytes32 contractId = _getContractId(contractType, contractName);
+
+        if (_contractRegistry[contractId] != address(0)) {
+            revert ContractAlreadyRegistered(contractType, contractName);
+        }
+
+        // Edge case: Cannot register a new "alias" (i.e., type/name) for an address if it's already been deprecated.
+        ContractInfo memory info = _contractInfo[contractAddress];
+        if (info.isRegistered && info.isActive == false) {
```

Comment:
So if it's registered _and_ active you can overwrite it with `register`? That sounds a bit awkward. Shouldn't you use `replace` for that? As far as I can tell, `replace` does allow the same address and a different name.

Reply:
There are some edge cases here. I think I can simplify it by 1) adding a deregister, so you can just delete and re-add, vs. applying complex replacement logic (should be very rare anyway); 2) disallowing having the same name with different types; I don't think that's a valid case, and complicates things; 3) making aliases explicit (probably a separate mapping), so that it ensures you're either adding a new pointer to an address, or changing the pointer to a different (valid) address. This prevents things like setting v2 and v3 of a contract to the same address (i.e., using a previously registered name as an "alias," creating orphan info entries).

Will convert to draft and work on it tomorrow.